src/vos/vos_io.c

```diff
@@ -1971,6 +1972,25 @@ vos_reserve_scm(struct vos_container *cont, struct vos_rsrvd_scm *rsrvd_scm,
 		umoff = umem_alloc(vos_cont2umm(cont), size);
 	}
 
+	vpm = cont->vc_pool->vp_metrics;
+	if (vpm && vpm->vp_space_metrics.vsm_scm_used) {
+		int		 rc;
+		struct vos_pool	*pool = vos_cont2pool(cont);
+		daos_size_t	 scm_used;
+
+		rc = pmemobj_ctl_get(pool->vp_umm.umm_pool,
+				     "stats.heap.run_allocated", &scm_used);
+		if (rc) {
+			rc = umem_tx_errno(rc);
+			D_ERROR("Query pool:"DF_UUID" SCM space failed. "DF_RC"\n",
+				DP_UUID(pool->vp_id), DP_RC(rc));
+			goto done;
+		}
+
+		d_tm_set_gauge(vpm->vp_space_metrics.vsm_scm_used, scm_used);
```

Comment:
The vos_reserve_scm() in only called for value allocation, so for the update request with >= 4k io size (value is landed to NVMe), the vos_reservec_scm() won't be called, but SCM will still be allocated for storing the key & interior tree node for this request, and if all incoming requests are large size updates, the SCM will be constantly consumed but the metric won't be updated.

I'm wondering if this 'scm_used' metric should be updated periodically by a ULT (like GC ULT)?

Reply:
Thanks for the explanation, which explain the inaccuracy of the SCM space used metrics.
From my side, I was thinking that the function `vos_reserve_scm()` was at least periodically called by the function `reserve_space()`  from the file _vos/vos_io.c_ as it was using a timeout constant `VOS_NOSPC_ERROR_INTVL`.

From my side, I already tried to insert the gauge update code in other functions periodically called from _src/vos/vos_aggregate.c_ and _src/vos/vos_pool_scrub.c_. Sadly, it was not working so well 😢 

As I am not familiar with this part of the code, could you tell more precisely, the ULT that you have in mind, which could be better adapted ?

Reply:
I think it's not necessary to create a dedicated ULT, you could put it in the GC ULT (in vos_gc_pool(), in the while loop and before the function exits).

Reply:
Initially, I was not sure which way was better: I was not sure if adding so side effect code was prohibited or not.
Happy to see that it is possible as it was little bit overkill for such task ;)
- [ ] Add metric update into GC ULT trhread.